### PR TITLE
Add tests for distriduted storage

### PR DESCRIPTION
**Pull request by AndreevSemen**

### CODE REVIEW

File: test/entrypoint/srv_kv_storage.lua [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        'app.roles.kv-storage',
+    },
+})
+
+if not ok then
+    log.error('%s', err)
+    os.exit(1)
+end
+
+_G.is_initialized = cartridge.is_healthy
```
Comment (dokshina) [inline test/entrypoint/srv_kv_storage.lua]: Please, add a newline.

File: test/helper.lua [hunk 2]
```diff
@@ -4,15 +4,21 @@
...
+end
 
-helper.root = fio.dirname(fio.abspath(package.search('init')))
-helper.datadir = fio.pathjoin(helper.root, 'tmp', 'db_test')
-helper.server_command = fio.pathjoin(helper.root, 'init.lua')
+local helpers = table.copy(cartridge_helpers)
+
+helpers.root = fio.dirname(fio.abspath(package.search('init')))
+helpers.datadir = fio.pathjoin(helpers.root, 'tmp', 'storage_test')
+helpers.server_command = fio.pathjoin(helpers.root, 'test', 'entrypoint', 'srv_kv_storage.lua')
```
Comment (dokshina) [inline test/helper.lua]: I think that it's not a good idea to have a single entrypoint and data directory for all tests.

I suggest using smth like https://github.com/tarantool/crud/blob/master/test/helper.lua#L36.

File: test/integration/api_test.lua [hunk 3]
```diff
@@ -1,12 +1,264 @@
...
-local cluster = helper.cluster
 
-g.test_sample = function()
-    local server = cluster.main_server
-    local response = server:http_request('post', '/admin/api', {json = {query = '{}'}})
-    t.assert_equals(response.json, {data = {}})
-    t.assert_equals(server.net_box:eval('return box.cfg.memtx_dir'), server.workdir)
+
+local http_client = require('http.client')
+local url = '/kv'
```
Comment (dokshina) [inline test/integration/api_test.lua]: Is it URL or endpoint?

File: test/integration/api_test.lua [hunk 4]
```diff
@@ -1,12 +1,264 @@
...
+    end
+end)
+
+g.after_each(function()
+
+end)
+
+-- CRUD helpers
+local function create(key, value)
+    return g.router:http_request('post', url..'/', {
```
Comment (dokshina) [inline test/integration/api_test.lua]: nit: I like `string.format('%s/', url)` more than `url .. '/'`

File: test/integration/api_test.lua [hunk 5]
```diff
@@ -1,12 +1,264 @@
...
+            key = key,
+            value = value,
+        },
+        http = { timeout = 1 },
+        raise = false,
+    })
+end
+
+local function read(key)
+    return g.router:http_request('get', url..'/'..key, { http = { timeout = 1 } , raise = false })
```
Comment (dokshina) [inline test/integration/api_test.lua]: The same as above.

BTW, aren't we have some "urljoin" functions anywhere in luatest?

File: test/integration/api_test.lua [hunk 6]
```diff
@@ -1,12 +1,264 @@
...
+end
+
+local function update_ok(key, value)
+    ok(update(key, value), { result = 'record updated' })
+end
+
+local function delete_ok(key)
+    ok(delete(key), { result = 'record deleted' })
+end
+
```
Comment (dokshina) [inline test/integration/api_test.lua]: I like this approach, but function names can be better:

*  `ok` -> `assert_resp_ok`
* `create_ok` - > `assert_create_ok`

and so on.

File: test/integration/api_test.lua [hunk 7]
```diff
@@ -1,12 +1,264 @@
...
+    not_found(delete('some_key'))
+end
+
+g.test_create_ok = function()
+    create_ok('awesome key', { some_json = { 'opt 1', 'opt 2', 'opt 3'} })
+end
+
+g.test_create_already_exists = function()
+    create_ok('some_key', {name = 'semen'})
+    create_already_exists('some_key', { array = {0, 1, 2, 3} })
```
Comment (dokshina) [inline test/integration/api_test.lua]: What do you think about smth like

```lua
local resp = create('some_key', {name = 'semen'})
assert_resp_ok(resp, <exp_body>)

local resp = create('some_key', {name = 'semen'})
assert_err_already_exists(resp)
```

It looks that `<*_not_found >` functions is too much, it's enough to separate functions for http requests, and functions to check response.

File: test/integration/api_test.lua (hunk 5 above)
Comment (AndreevSemen) [reply test/integration/api_test.lua]: I searched but couldn't find std way to join url. So I wrote my own

File: test/helper/integration.lua [hunk 8]
```diff
@@ -5,4 +5,28 @@ local fio = require('fio')
...
+        if i == 1 then
+            url = piece
+        else
+            url = string.format('%s/%s', url, piece)
+        end
+    end
+    return url
+end
+
+function helper.path_to_storage(storage)
```
Comment (dokshina) [inline test/helper/integration.lua]: What is `storage`? It looks like `storage_name`.

Moreover, I think that you can define `/kv` as a `base_endpoint` or smth like this in the test file directly (or `kv_base_endpoint` in helper)

File: test/integration/api_test.lua [hunk 9]
```diff
@@ -1,22 +1,22 @@
...
 
+local helper = require('test.helper.integration')
 
-local http_client = require('http.client')
-local url = '/kv'
+local server_command = 'srv_kv_storage'
 
+local storage_name = 'kv'
+local storage_path = helper.path_to_storage(storage_name)
+local path_to_key = helper.key_path_formatter(storage_path)
```
Comment (dokshina) [inline test/integration/api_test.lua]: nit: think about naming - `path_to_key` looks like a path (some `string`), but it's a function.

```lua
local key_path_formatter = helper.get_key_path_formatter(endpoint)
local path = key_path_formatter(key)
```

or simply

```lua
local path = helper.format_key_path(endpoint, key)
```

File: test/helper/integration.lua (hunk 8 above)
Comment (AndreevSemen) [reply test/helper/integration.lua]: Yeah, really, there are no any storages are expected. So I can make common `base_kv_enpoint` for all of tests

File: test/integration/api_test.lua (hunk 9 above)
Comment (AndreevSemen) [reply test/integration/api_test.lua]: I made common `base_kv_enpoint` from previous conversation, so it's good to make helper `key_path(key)` (anyway it's a singleton):

```lua
-- in helper/integration.lua
function helper.key_path(key)
    return join_url(helper.base_kv_endpoint, key)
end
```

**Review by dokshina — APPROVED** [review]

**Pull request merged**